### ISSUE TITLE
Task-Free Continual Learning

### ISSUE THREAD

**Issue by juanmc2005** [juanmc2005/phd-bib#28]
## [Task-Free Continual Learning](http://openaccess.thecvf.com/content_CVPR_2019/html/Aljundi_Task-Free_Continual_Learning_CVPR_2019_paper.html)
_Methods proposed in the literature towards continual deep learning typically operate in a task-based sequential learning setup. A sequence of tasks is learned, one at a time, with all data of current task available but not of previous or future tasks. Task boundaries and identities are known at all times. This setup, however, is rarely encountered in practical applications. Therefore we investigate how to transform continual learning to an online setup. We develop a system that keeps on learning over time in a streaming fashion, with data distributions gradually changing and without the notion of separate tasks. To this end, we build on the work on Memory Aware Synapses, and show how this method can be made online by providing a protocol to decide i) when to update the importance weights, ii) which data to use to update them, and iii) how to accumulate the importance weights at each update step. Experimental results show the validity of the approach in the context of two applications: (self-)supervised learning of a face recognition model by watching soap series and learning a robot to avoid collisions._

### Cite this paper
```bibtex
@InProceedings{Aljundi_2019_CVPR,
    author = {Aljundi, Rahaf and Kelchtermans, Klaas and Tuytelaars, Tinne},
    title = {Task-Free Continual Learning},
    booktitle = {The IEEE Conference on Computer Vision and Pattern Recognition (CVPR)},
    month = {June},
    year = {2019}
}
```

**Comment by juanmc2005** [comment]
### Description
This paper argues that current continual learning methods are limited, in the fact that task boundaries are known beforehand, while this is rarely the case in real world scenarios.
They propose a learning algorithm and architecture based on Memory Aware Synapses (MAS), that assigns importance scores to weights to avoid catastrophic interference of consolidated knowledge.
They test their method on three different scenarios: a synthetic dataset, face recognition on TV series and collision avoidance with a robot. I'll focus on the second one in this summary because it is the most related to our use case.

### Method
At each time step, the model receives a batch of consecutive samples using a weak supervised or self-supervised signal.

As task boundaries are not known, the distribution of the input data might change gradually, so they need to optimize an objective while being able to shift to this new distributions without forgetting previous ones, hence accumulating knowledge. They calculate the total loss of the model using a replay method, in which they store the hardest N examples to the current time step. The loss is then calculated as the loss over the current sequence plus the loss over the hardest examples stored.

However, the system is still vulnerable to catastrophic forgetting, as the sequential nature of the data and the small number of samples can still bias the model towards a specific distribution. This is why they introduce a variant of MAS, which estimates the importance of the weights after learning each task. As tasks are not known in this setting, they estimate distribution changes by looking at the loss with a sliding window and calculating the mean and variance. If the model observes a plateau, then the distribution hasn't shifted, but if there's a peak, then that means the data distribution has changed considerably for the model to make big errors.

![image](https://user-images.githubusercontent.com/5760780/77057143-cc033680-69d3-11ea-9a1a-e1fcbef00510.png)

### Experiments
Face recognition experiments are conducted using an AlexNet pretrained on ImageNet, on episodes of The Big Bang Theory, Breaking Bad and Mad Men. They optimize a triplet loss on weakly supervised and self-supervised settings. Their baselines include:
- Initial: the initial pretrained model
- Online Baseline: a continual model without MAS
- Online Joint Training: a continual model without MAS trained on shuffled data
- Offline Joint Training: a non-continual model trained on shuffled data for multiple epochs, which constitutes an upper bound 

### Results
#### Weak Supervision
Using a weak supervised signal (knowing if 2 faces from consecutive batches are the same), they observe that the Online Baseline performs better than the Initial model but then fails to accumulate knowledge, while Online Joint and their proposed method consolidate this knowledge better. However, when changing the distribution drastically (changing series), they observe a clear advantage of their model, even coming close to the upper bound.

#### Self-Supervision
Using a self-supervised signal (knowing that 2 faces in the same frame belong to different identities), they observe similar results, with their method obtaining better performance over time, compared to the Online Baseline and the Initial model. However, it is not clear why they didn't test in this setting the other baselines they proposed.

### Takeaway
The most interesting takeaway of this paper is that they tackle the challenging setting of not having task boundaries, which in my view constitutes a real continual learning scenario. Additionally, they have trained their models with triplet loss and evaluation consisted of a nearest neighbor algorithm to determine the identity of a face. This shows that metric learning losses we know can be used not only in a continual learning setting but also in a self-supervised one, which is encouraging for my PhD subject.
